File: backend/src/routes/cloud-usage.ts
import { Hono } from 'hono';
import { promises as fs } from 'fs';
import path from 'path';
import { getDataDir } from '../utils/helpers';

const USAGE_FILE = path.join(getDataDir(), 'cloud_usage.json');

// ─── Model Pricing Table (USD per 1M tokens) ──────────────────────
// Covers the most common cloud models. Users can override via custom pricing.
const DEFAULT_PRICING: Record<string, { input: number; output: number }> = {
  // OpenAI
  'gpt-4o':            { input: 2.50,  output: 10.00 },
  'gpt-4o-mini':       { input: 0.15,  output: 0.60 },
  'gpt-4-turbo':       { input: 10.00, output: 30.00 },
  'gpt-4':             { input: 30.00, output: 60.00 },
  'gpt-3.5-turbo':     { input: 0.50,  output: 1.50 },
  'o1':                { input: 15.00, output: 60.00 },
  'o1-mini':           { input: 3.00,  output: 12.00 },
  'o3-mini':           { input: 1.10,  output: 4.40 },
  // Anthropic
  'claude-sonnet-4-20250514':  { input: 3.00,  output: 15.00 },
  'claude-3-5-sonnet-20241022':{ input: 3.00,  output: 15.00 },
  'claude-3-5-haiku-20241022': { input: 0.80,  output: 4.00 },
  'claude-3-opus-20240229':    { input: 15.00, output: 75.00 },
  'claude-3-haiku-20240307':   { input: 0.25,  output: 1.25 },
  // Google
  'gemini-2.0-flash':          { input: 0.10,  output: 0.40 },
  'gemini-1.5-pro':            { input: 1.25,  output: 5.00 },
  'gemini-1.5-flash':          { input: 0.075, output: 0.30 },
  // DeepSeek
  'deepseek-chat':      { input: 0.14,  output: 0.28 },
  'deepseek-coder':     { input: 0.14,  output: 0.28 },
  'deepseek-reasoner':  { input: 0.55,  output: 2.19 },
};

interface CloudUsage {
  /** Total cloud API requests made since last reset */
  totalRequests: number;
  /** Estimated total tokens consumed */
  totalTokens: number;
  /** Requests per model */
  byModel: Record<string, number>;
  /** Per-model token tracking: { model: { input, output } } */
  tokensByModel: Record<string, { input: number; output: number }>;
  /** Monthly limit (number of requests) — 0 = unlimited */
  monthlyLimit: number;
  /** Token budget — 0 = unlimited */
  tokenBudget: number;
  /** Custom pricing overrides: { model: { input, output } } per 1M tokens */
  customPricing: Record<string, { input: number; output: number }>;
  /** ISO date of last reset */
  lastReset: string;
  /** ISO date of current tracking period start (month boundary) */
  periodStart: string;
}

interface CostBreakdown {
  model: string;
  inputTokens: number;
  outputTokens: number;
  inputCost: number;
  outputCost: number;
  totalCost: number;
}

const DEFAULT_USAGE: CloudUsage = {
  totalRequests: 0,
  totalTokens: 0,
  byModel: {},
  tokensByModel: {},
  monthlyLimit: 0,
  tokenBudget: 0,
  customPricing: {},
  lastReset: new Date().toISOString(),
  periodStart: getMonthStart(),
};

function getMonthStart(): string {
  const now = new Date();
  return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-01`;
}

function isNewMonth(usage: CloudUsage): boolean {
  return usage.periodStart !== getMonthStart();
}

async function ensureDir(): Promise<void> {
  await fs.mkdir(getDataDir(), { recursive: true });
}

async function loadUsage(): Promise<CloudUsage> {
  try {
    await ensureDir();
    const data = await fs.readFile(USAGE_FILE, 'utf-8');
    const parsed = JSON.parse(data) as CloudUsage;
    // Auto-reset if a new month started
    if (isNewMonth(parsed)) {
      parsed.totalRequests = 0;
      parsed.totalTokens = 0;
      parsed.byModel = {};
      parsed.tokensByModel = {};
      parsed.periodStart = getMonthStart();
      parsed.lastReset = new Date().toISOString();
      await saveUsage(parsed);
    }
    return { ...DEFAULT_USAGE, ...parsed };
  } catch {
    return DEFAULT_USAGE;
  }
}

async function saveUsage(usage: CloudUsage): Promise<void> {
  await ensureDir();
  await fs.writeFile(USAGE_FILE, JSON.stringify(usage, null, 2));
}

/**
 * Get pricing for a model — custom overrides take priority, then fuzzy match
 * against the default table (prefix match for provider-prefixed model names).
 */
function getModelPricing(
  model: string,
  customPricing: Record<string, { input: number; output: number }>
): { input: number; output: number } | null {
  // 1. Exact custom override
  if (customPricing[model]) return customPricing[model];
  // 2. Exact default
  if (DEFAULT_PRICING[model]) return DEFAULT_PRICING[model];
  // 3. Fuzzy: model contains a known key (e.g. "openai/gpt-4o" → "gpt-4o")
  const lower = model.toLowerCase();
  for (const [key, price] of Object.entries(DEFAULT_PRICING)) {
    if (lower.includes(key)) return price;
  }
  // 4. Custom pricing prefix match
  for (const [key, price] of Object.entries(customPricing)) {
    if (lower.includes(key.toLowerCase())) return price;
  }
  return null;
}

/**
 * Calculate estimated cost for a model given input/output token counts.
 */
function calculateModelCost(
  model: string,
  inputTokens: number,
  outputTokens: number,
  customPricing: Record<string, { input: number; output: number }>
): number {
  const pricing = getModelPricing(model, customPricing);
  if (!pricing) return 0;
  // Pricing is per 1M tokens
  const inputCost = (inputTokens / 1_000_000) * pricing.input;
  const outputCost = (outputTokens / 1_000_000) * pricing.output;
  return inputCost + outputCost;
}

const cloudUsage = new Hono();

// GET / — read current usage with cost breakdown
cloudUsage.get('/', async (c) => {
  const usage = await loadUsage();

  // Build cost breakdown per model
  const breakdown: CostBreakdown[] = [];
  let totalEstimatedCost = 0;

  for (const [model, tokens] of Object.entries(usage.tokensByModel)) {
    const inputTokens = tokens.input || 0;
    const outputTokens = tokens.output || 0;
    const pricing = getModelPricing(model, usage.customPricing);

    let inputCost = 0;
    let outputCost = 0;
    if (pricing) {
      inputCost = (inputTokens / 1_000_000) * pricing.input;
      outputCost = (outputTokens / 1_000_000) * pricing.output;
    }
    const totalCost = inputCost + outputCost;
    totalEstimatedCost += totalCost;

    breakdown.push({
      model,
      inputTokens,
      outputTokens,
      inputCost,
      outputCost,
      totalCost,
    });
  }

  // Sort breakdown by cost descending
  breakdown.sort((a, b) => b.totalCost - a.totalCost);

  // Pricing info for known models
  const pricingTable: Record<string, { input: number; output: number; source: string }> = {};
  for (const [model, price] of Object.entries(DEFAULT_PRICING)) {
    pricingTable[model] = { ...price, source: 'built-in' };
  }
  for (const [model, price] of Object.entries(usage.customPricing)) {
    pricingTable[model] = { ...price, source: 'custom' };
  }

  return c.json({
    ...usage,
    costBreakdown: breakdown,
    totalEstimatedCost,
    pricingTable,
  });
});

// POST /increment — called by the pipeline when a cloud request is made
cloudUsage.post('/increment', async (c) => {
  try {
    const body = await c.req.json();
    const { model, tokens, inputTokens, outputTokens } = body as {
      model?: string;
      tokens?: number;
      inputTokens?: number;
      outputTokens?: number;
    };
    const usage = await loadUsage();
    usage.totalRequests += 1;

    // Token tracking (legacy flat total)
    const tokenDelta = tokens || 0;
    usage.totalTokens += tokenDelta;

    // Per-model request count
    if (model) {
      usage.byModel[model] = (usage.byModel[model] || 0) + 1;

      // Per-model token tracking (input/output split)
      if (!usage.tokensByModel) usage.tokensByModel = {};
      if (!usage.tokensByModel[model]) usage.tokensByModel[model] = { input: 0, output: 0 };

      if (inputTokens !== undefined || outputTokens !== undefined) {
        // Split tokens provided
        usage.tokensByModel[model].input += inputTokens || 0;
        usage.tokensByModel[model].output += outputTokens || 0;
      } else if (tokenDelta > 0) {
        // Legacy: approximate 60/40 input/output split
        usage.tokensByModel[model].input += Math.round(tokenDelta * 0.6);
        usage.tokensByModel[model].output += Math.round(tokenDelta * 0.4);
      }
    }

    await saveUsage(usage);
    return c.json(usage);
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : 'Failed to increment' }, 500);
  }
});

// PUT /limit — update the monthly request limit and token budget
cloudUsage.put('/limit', async (c) => {
  try {
    const body = await c.req.json();
    const usage = await loadUsage();
    if (body.monthlyLimit !== undefined) usage.monthlyLimit = Number(body.monthlyLimit) || 0;
    if (body.tokenBudget !== undefined) usage.tokenBudget = Number(body.tokenBudget) || 0;
    await saveUsage(usage);
    return c.json(usage);
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : 'Failed to update limit' }, 500);
  }
});

// PUT /pricing — update custom pricing overrides
cloudUsage.put('/pricing', async (c) => {
  try {
    const body = await c.req.json();
    const { model, input, output } = body as { model?: string; input?: number; output?: number };
    if (!model) return c.json({ error: 'model is required' }, 400);
    const usage = await loadUsage();
    if (!usage.customPricing) usage.customPricing = {};
    if (input !== undefined && output !== undefined) {
      usage.customPricing[model] = { input, output };
    } else {
      // Remove custom pricing (revert to built-in)
      delete usage.customPricing[model];
    }
    await saveUsage(usage);
    return c.json({ customPricing: usage.customPricing });
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : 'Failed to update pricing' }, 500);
  }
});

// POST /reset — reset counters (keep limits and custom pricing)
cloudUsage.post('/reset', async (c) => {
  const usage = await loadUsage();
  usage.totalRequests = 0;
  usage.totalTokens = 0;
  usage.byModel = {};
  usage.tokensByModel = {};
  usage.lastReset = new Date().toISOString();
  usage.periodStart = getMonthStart();
  await saveUsage(usage);
  return c.json(usage);
});

export default cloudUsage;

File: backend/app/routes/cloud_usage.py
"""Cloud usage tracking routes — mirrors backend/src/routes/cloud-usage.ts."""

from __future__ import annotations

import json
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def _js_num(v: Any) -> Any:
    """Match JS JSON.stringify: whole floats serialize as integers (4.00 → 4)."""
    if isinstance(v, float) and v.is_integer():
        return int(v)
    return v


def _js_iso(dt: datetime | None = None) -> str:
    """Format like JS Date.toISOString(): UTC, milliseconds, Z suffix.

    The TS backend emits lastReset this way and the frontend parses it, so the
    Python port must match exactly (no microseconds, no +00:00 offset).
    """
    d = (dt or datetime.now(timezone.utc)).astimezone(timezone.utc)
    return d.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..config import get_data_dir

router = APIRouter()

USAGE_FILE = Path(get_data_dir()) / "cloud_usage.json"

# ─── Model Pricing Table (USD per 1M tokens) ──────────────────────
DEFAULT_PRICING: dict[str, dict[str, float]] = {
    "gpt-4o": {"input": 2.50, "output": 10.00},
    "gpt-4o-mini": {"input": 0.15, "output": 0.60},
    "gpt-4-turbo": {"input": 10.00, "output": 30.00},
    "gpt-4": {"input": 30.00, "output": 60.00},
    "gpt-3.5-turbo": {"input": 0.50, "output": 1.50},
    "o1": {"input": 15.00, "output": 60.00},
    "o1-mini": {"input": 3.00, "output": 12.00},
    "o3-mini": {"input": 1.10, "output": 4.40},
    "claude-sonnet-4-20250514": {"input": 3.00, "output": 15.00},
    "claude-3-5-sonnet-20241022": {"input": 3.00, "output": 15.00},
    "claude-3-5-haiku-20241022": {"input": 0.80, "output": 4.00},
    "claude-3-opus-20240229": {"input": 15.00, "output": 75.00},
    "claude-3-haiku-20240307": {"input": 0.25, "output": 1.25},
    "gemini-2.0-flash": {"input": 0.10, "output": 0.40},
    "gemini-1.5-pro": {"input": 1.25, "output": 5.00},
    "gemini-1.5-flash": {"input": 0.075, "output": 0.30},
    "deepseek-chat": {"input": 0.14, "output": 0.28},
    "deepseek-coder": {"input": 0.14, "output": 0.28},
    "deepseek-reasoner": {"input": 0.55, "output": 2.19},
}


def _get_month_start() -> str:
    now = datetime.now()
    return f"{now.year}-{now.month:02d}-01"


def _is_new_month(usage: dict[str, Any]) -> bool:
    return usage.get("periodStart") != _get_month_start()


def _default_usage() -> dict[str, Any]:
    return {
        "totalRequests": 0,
        "totalTokens": 0,
        "byModel": {},
        "tokensByModel": {},
        "monthlyLimit": 0,
        "tokenBudget": 0,
        "customPricing": {},
        "lastReset": _js_iso(),
        "periodStart": _get_month_start(),
    }


async def load_usage() -> dict[str, Any]:
    try:
        parsed = json.loads(USAGE_FILE.read_text(encoding="utf-8"))
        usage = {**_default_usage(), **parsed}
        if _is_new_month(usage):
            usage["totalRequests"] = 0
            usage["totalTokens"] = 0
            usage["byModel"] = {}
            usage["tokensByModel"] = {}
            usage["periodStart"] = _get_month_start()
            usage["lastReset"] = _js_iso()
            await save_usage(usage)
        return usage
    except (OSError, json.JSONDecodeError):
        return _default_usage()


async def save_usage(usage: dict[str, Any]) -> None:
    USAGE_FILE.parent.mkdir(parents=True, exist_ok=True)
    USAGE_FILE.write_text(json.dumps(usage, indent=2), encoding="utf-8")


def _get_model_pricing(model: str, custom_pricing: dict[str, dict[str, float]]) -> dict[str, float] | None:
    if model in custom_pricing:
        return custom_pricing[model]
    if model in DEFAULT_PRICING:
        return DEFAULT_PRICING[model]
    lower = model.lower()
    for key, price in DEFAULT_PRICING.items():
        if key in lower:
            return price
    for key, price in custom_pricing.items():
        if key.lower() in lower:
            return price
    return None


def _calculate_model_cost(model: str, input_tokens: int, output_tokens: int, custom_pricing: dict) -> float:
    pricing = _get_model_pricing(model, custom_pricing)
    if not pricing:
        return 0.0
    return (input_tokens / 1_000_000) * pricing["input"] + (output_tokens / 1_000_000) * pricing["output"]


@router.get("")
async def get_usage() -> dict:
    usage = await load_usage()
    breakdown: list[dict[str, Any]] = []
    total_estimated_cost = 0.0
    for model, tokens in usage.get("tokensByModel", {}).items():
        input_tokens = tokens.get("input") or 0
        output_tokens = tokens.get("output") or 0
        pricing = _get_model_pricing(model, usage.get("customPricing", {}))
        input_cost = (input_tokens / 1_000_000) * pricing["input"] if pricing else 0.0
        output_cost = (output_tokens / 1_000_000) * pricing["output"] if pricing else 0.0
        total_cost = input_cost + output_cost
        total_estimated_cost += total_cost
        breakdown.append(
            {
                "model": model,
                "inputTokens": input_tokens,
                "outputTokens": output_tokens,
                "inputCost": input_cost,
                "outputCost": output_cost,
                "totalCost": total_cost,
            }
        )
    breakdown.sort(key=lambda b: b["totalCost"], reverse=True)
    pricing_table: dict[str, Any] = {}
    for model, price in DEFAULT_PRICING.items():
        pricing_table[model] = {"input": _js_num(price["input"]), "output": _js_num(price["output"]), "source": "built-in"}
    for model, price in usage.get("customPricing", {}).items():
        pricing_table[model] = {"input": _js_num(price.get("input")), "output": _js_num(price.get("output")), "source": "custom"}
    return {
        **usage,
        "costBreakdown": breakdown,
        "totalEstimatedCost": _js_num(total_estimated_cost),
        "pricingTable": pricing_table,
    }


@router.post("/increment")
async def increment(request: Request) -> dict:
    try:
        body = await request.json()
        model = body.get("model")
        tokens = body.get("tokens")
        input_tokens = body.get("inputTokens")
        output_tokens = body.get("outputTokens")
        usage = await load_usage()
        usage["totalRequests"] += 1
        token_delta = tokens or 0
        usage["totalTokens"] += token_delta
        if model:
            usage.setdefault("byModel", {})
            usage["byModel"][model] = usage["byModel"].get(model, 0) + 1
            usage.setdefault("tokensByModel", {})
            usage["tokensByModel"].setdefault(model, {"input": 0, "output": 0})
            if input_tokens is not None or output_tokens is not None:
                usage["tokensByModel"][model]["input"] += input_tokens or 0
                usage["tokensByModel"][model]["output"] += output_tokens or 0
            elif token_delta > 0:
                usage["tokensByModel"][model]["input"] += round(token_delta * 0.6)
                usage["tokensByModel"][model]["output"] += round(token_delta * 0.4)
        await save_usage(usage)
        return usage
    except Exception as e:  # noqa: BLE001
        return JSONResponse({"error": str(e) or "Failed to increment"}, status_code=500)


@router.put("/limit")
async def update_limit(request: Request) -> dict:
    try:
        body = await request.json()
        usage = await load_usage()
        if body.get("monthlyLimit") is not None:
            usage["monthlyLimit"] = float(body["monthlyLimit"]) or 0
        if body.get("tokenBudget") is not None:
            usage["tokenBudget"] = float(body["tokenBudget"]) or 0
        await save_usage(usage)
        return usage
    except Exception as e:  # noqa: BLE001
        return JSONResponse({"error": str(e) or "Failed to update limit"}, status_code=500)


@router.put("/pricing")
async def update_pricing(request: Request) -> dict:
    try:
        body = await request.json()
        model = body.get("model")
        if not model:
            return JSONResponse({"error": "model is required"}, status_code=400)
        usage = await load_usage()
        usage.setdefault("customPricing", {})
        if body.get("input") is not None and body.get("output") is not None:
            usage["customPricing"][model] = {"input": body["input"], "output": body["output"]}
        else:
            usage["customPricing"].pop(model, None)
        await save_usage(usage)
        return {"customPricing": usage["customPricing"]}
    except Exception as e:  # noqa: BLE001
        return JSONResponse({"error": str(e) or "Failed to update pricing"}, status_code=500)


@router.post("/reset")
async def reset_usage() -> dict:
    usage = await load_usage()
    usage["totalRequests"] = 0
    usage["totalTokens"] = 0
    usage["byModel"] = {}
    usage["tokensByModel"] = {}
    usage["lastReset"] = _js_iso()
    usage["periodStart"] = _get_month_start()
    await save_usage(usage)
    return usage